Disable start button in frontend until backend becomes available

rohanpurohit opened the pull request:
… (#325)

rohanpurohit committed:
frontend: disable start button landing page if backend is unavailable (#325)

ethanjli commented:
Right now the behavior I observe on the landing page is:

- If I start the frontend without the backend running, the landing page briefly flashes "Start" and then (within ~500 ms, based on the timeout you set) switches to "Loading"
- If I start the frontend after the backend is running, the landing page briefly flashes "Start", then changes to "Loading", then switches back to "Start"

I see this behavior on the landing page, but not on the quickstart screen.

I believe the reason this behavior happens is that you assume that the backend is connected at the start, then you wait to see if the backend is lost based on a timeout which fires "BACKEND_CONNECTION_LOST". Maybe it would be cleaner if the button started as "Loading" by default, and then a "BACKEND_CONNECTION_INITIALIZED" event were fired which changed it to "Start"?

Minor feedback: let's have the button say "Loading...", rather than just "Loading".

ethanjli commented:
Note related to futureproofing for the first task I'm working on this week (making the frontend initialize its store's ParametersRequest, AlarmLimitsRequest, etc., as `undefined` and wait for the backend to give it the initial values): (ignore everything in italics, we discussed a better solution in our meeting) _eventually we will want BACKEND_CONNECTION_INITIALIZED to be dispatched once the store's ParametersRequest, AlarmLimitsRequest, etc., are no longer undefined. It would be helpful if you could create an action for BACKEND_CONNECTION_INITIALIZED in such a way that I can go back in and modify its trigger conditions, e.g. using a selector._ The approach we discussed trying out is just to make a selector which decides whether the backend connection is initialized, based on the values of things in the store.

rohanpurohit committed:
update implementation after store changes

add changes for backend initialized or disconnected in store

update landing page stuff

update toolbar logic

rohanpurohit committed:
Merge branch 'develop' into hotfix/frontend-landing-screen-start

rohanpurohit commented:
Still working on it! has known issues, pushed to discuss approach and code review

rohanpurohit committed:
reimplement landing page and toolbar logic

rohanpurohit commented:
TODO: 
- add EventLog for backend_connection_up , also change naming for `backend_connection_down `which is currently `backend_connection_lost_code` 
- figure out a robust way to handle toolbar logic for `start/pause` button (problem: no robust way of figuring out if we are on landing page, currently we just send a boolean if start button is clicked on landing page which is `staticStart`, but we need to figure out that we are on landing page before that)
- useEffect that handles dispatching of `BACKEND_CONNECTION_DOWN `and `BACKEND_CONNECTION_UP` needs more refactoring?, does not work if it's in a single `useEffect` with an if-else block (maybe because it is only scheduled once?)
- there is some race condition on START and Pause ventilation from quickstart to dashboard (linked to toolbar logic)

@ethanjli  any thoughts overall?

rohanpurohit committed:
Fix toolbar logic, dispatch log event for backend connection down

rohanpurohit committed:
fix label on landing screen

rohanpurohit committed:
trivial fixes

rohanpurohit commented:
@ethanjli This PR is ready for testing and code review.

rohanpurohit committed:
frontend: change loading behavior

rohanpurohit committed:
toolbar: fix typo for useState

ethanjli commented on frontend/src/store/app/actions.ts:
```diff
@@ -7,3 +13,11 @@ export function initialize(): AppAction {
 }
+
+export function lostBackendConnection(backendConnection: boolean, clock: Date): AppAction {
+  return { type: BACKEND_CONNECTION_DOWN, clock, backendConnection };
```
Do we actually Need the "backendConnection" boolean field in these actions? What do the following actions mean?
```
{
  type: BACKEND_CONNECTION_DOWN,
  backendConnection: true
}
```
```
{
  type: BACKEND_CONNECTION_DOWN,
  backendConnection: false
}
```

(this also applies to the definitions of the `BackendConnectionDown` and `BackendConnectionUp` interfaces in `types.ts`)

ethanjli commented on frontend/src/store/app/reducers.ts:
```diff
@@ -31,6 +34,10 @@ export function appReducer(state = initialState, action: AppAction): AppState {
       // it should come from the code which creates the action to dispatch.
       return { ...state, backendHeartbeat: new Date() };
+    case BACKEND_CONNECTION_DOWN:
+      return { ...state, clock: new Date(), backendConnection: action.backendConnection };
```
Could we just set `backendConnection: false` for `BACKEND_CONNECTION_DOWN` and `backendConnection: true` for BACKEND_CONNECTION_UP`?

ethanjli commented on frontend/src/store/app/selectors.ts:
```diff
@@ -31,6 +31,14 @@ export const getAlarmNotifyStatus = createSelector(
+);
+export const getBackendDisconnected = createSelector(
+  getApp,
+  (app: AppState): boolean => app.backendConnection,
```
From naming, I would assume that `app.backendConnection` represents whether the backend is connected, while a `getBackendDisconnected` selector represents the opposite of that - i.e. shouldn't `getBackendDisconnected` return `!app.backendConnection`?

ethanjli commented on frontend/src/store/controller/selectors.ts:
```diff
@@ -233,6 +234,14 @@ export const getPopupEventLog = createSelector(getController, (states: Controlle
+  getSensorMeasurements,
+  getBackendDisconnected,
+  (sensorMeasurements: SensorMeasurements | null, backendDisconnected: boolean): boolean =>
+    sensorMeasurements !== null && backendDisconnected,
```
Based on function names, the logic expressed in this selector is that the backend is initialized if both of the following conditions are true:
- sensorMeasurements is not-null
- the backend is disconnected.
Based on this, and how you implemented the `getBackendDisconnected` selector, maybe you actually meant to name the selector `getBackendConnected` rather than `getBackendDisconnected`?

ethanjli replied:
While we're at it, let's change `sensorMeasurements !== null` to `parametersRequest !== null && alarmLimitsRequest !== null` in this PR.

ethanjli commented on frontend/src/modules/app/OverlayScreen.tsx:
```diff
@@ -36,34 +37,41 @@ const useStyles = makeStyles((theme: Theme) => ({
+  const events = useSelector(getNextLogEvents);
+  const diff = Math.abs(new Date().valueOf() - new Date(heartbeat).valueOf());
+
+  useEffect(() => {
+    if (diff < BACKEND_CONNECTION_TIMEOUT) {
+      dispatch(establishedBackendConnection(true, new Date()));
+    }
+  }, [clock, diff, dispatch, heartbeat]);
 
   useEffect(() => {
```
Since these just cover different branches of the same `if` check (i.e. `diff <= BACKEND_CONNECTION_TIMEOUT`), would it make sense to merge these two into the same `useEffect`? Like:
```
if (diff <= BACKEND_CONNECTION_TIMEOUT {
  ...
} else if (!lostConnectionAlarm) {
  ...
}
```

ethanjli reviewed (commented):
It works in basic testing, here's a review of the code.

rohanpurohit committed:
frontend: resolve review comments

rohanpurohit commented:
1. This project is licensed under Apache License v2.0 for any software, and Solderpad Hardware License v2.1 for any hardware - do you agree that your contributions to this project will be under these licenses, too? Yes.
2. Were any of these contributions also part of work you did for an employer or a client? No.
3. Does this work include, or is it based on, any third-party work which you did not create? No.

ethanjli reviewed (approved):
I tested the new changes, it works well!
